File: docs/JAMIA/comparisons/RPython/TM/17151210_loadtxt_header_row.ts
/**
 * RPython SO#17151210 — numpy loadtxt skip first row
 * Effect: Crash
 * Bug class: Data loading
 *
 * Python bug: CSV has comment lines (#) and a header row (x,y,z). User uses
 * np.loadtxt(comments='#', skiprows=1) but the header row "x,y,z" is not a
 * comment and not skipped enough — numpy crashes: "could not convert string to
 * float: x". The non-numeric header was included in the numeric parse.
 *
 * In tidy-ts, readCSV with a Zod schema validates at runtime (z.number() rejects
 * "x" at parse). At the type level, if the schema uses z.string() (wrong schema),
 * downstream numeric operations on the resulting string columns are rejected.
 */
import { readCSV, stats as s } from "@tidy-ts/dataframe";
import { z } from "zod";

// If loaded with wrong schema (string instead of number), the DataFrame
// columns are typed as string — downstream numeric ops are rejected.
const wrongSchema = z.object({ x: z.string(), y: z.string(), z: z.string() });

readCSV("x,y,z\n1,2,3\n4,5,6", wrongSchema).then((df) => {
  // @ts-expect-error — string[] is not assignable to number[]
  s.mean(df.extract("x"));
});

File: docs/JAMIA/comparisons/RPython/TM/17151210_loadtxt_header_row.py
# ID: SO#17151210
# Language: Python
# Bug class: Data loading
# Runtime consequence: Crash
# In study: Yes
# Inclusion rationale: numpy loadtxt fails on header/comment rows. Non-numeric content in numeric load.
# Reproduction status: Live
# Type system catch: `Record<string, number>` rejects string column

import numpy as np
import tempfile
import os

# Create the CSV file matching the SO scenario
content = """# Comment 1
# Comment 2
x,y,z
1,2,3
4,5,6
7,8,9
"""

tmpfile = tempfile.NamedTemporaryFile(mode='w', suffix='.csv', delete=False)
tmpfile.write(content)
tmpfile.close()

# The SO bug: skiprows=1 only skips 1 line after comments, but the header
# row "x,y,z" is NOT a comment (no #), so it gets included in numeric parse
try:
    FH = np.loadtxt(tmpfile.name, comments='#', delimiter=',', skiprows=1)
    print("Unexpectedly succeeded")
except ValueError as e:
    print(f"Crash: ValueError: {e}")

print()

# The fix: skip the header row properly (skiprows=2 or filter non-comment headers)
with open(tmpfile.name) as f:
    lines = (line for line in f if not line.startswith('#'))
    FH = np.loadtxt(lines, delimiter=',', skiprows=1)
print("With proper header skipping:")
print(FH)

os.unlink(tmpfile.name)
